source/common/http/async_client_impl.cc

```diff
@@ -22,120 +21,182 @@ AsyncClientImpl::AsyncClientImpl(const Upstream::ClusterInfo& cluster, Stats::St
+  AsyncRequestImpl* async_request =
+      new AsyncRequestImpl(std::move(request), *this, callbacks, timeout);
+  std::unique_ptr<AsyncStreamImpl> new_request{async_request};
 
   // The request may get immediately failed. If so, we will return nullptr.
-  if (!new_request->complete_) {
-    new_request->moveIntoList(std::move(new_request), active_requests_);
-    return active_requests_.front().get();
+  if (!new_request->complete()) {
+    new_request->moveIntoList(std::move(new_request), active_streams_);
+    return async_request;
   } else {
     return nullptr;
   }
 }
 
-AsyncRequestImpl::AsyncRequestImpl(MessagePtr&& request, AsyncClientImpl& parent,
-                                   AsyncClient::Callbacks& callbacks,
-                                   const Optional<std::chrono::milliseconds>& timeout)
-    : request_(std::move(request)), parent_(parent), callbacks_(callbacks),
-      stream_id_(parent.config_.random_.random()), router_(parent.config_),
-      request_info_(Protocol::Http11), route_(parent_.cluster_.name(), timeout) {
+AsyncClient::Stream* AsyncClientImpl::start(AsyncClient::StreamCallbacks& callbacks,
+                                            const Optional<std::chrono::milliseconds>& timeout) {
+  std::unique_ptr<AsyncStreamImpl> new_stream{new AsyncStreamImpl(*this, callbacks, timeout)};
 
-  router_.setDecoderFilterCallbacks(*this);
-  request_->headers().insertEnvoyInternalRequest().value(
-      Headers::get().EnvoyInternalRequestValues.True);
-  request_->headers().insertForwardedFor().value(parent_.config_.local_info_.address());
-  router_.decodeHeaders(request_->headers(), !request_->body());
-  if (!complete_ && request_->body()) {
-    router_.decodeData(*request_->body(), true);
+  // The request may get immediately failed. If so, we will return nullptr.
+  if (!new_stream->complete()) {
+    new_stream->moveIntoList(std::move(new_stream), active_streams_);
+    return active_streams_.front().get();
+  } else {
+    return nullptr;
   }
+}
 
-  // TODO: Support request trailers.
+AsyncStreamImpl::AsyncStreamImpl(AsyncClientImpl& parent, AsyncClient::StreamCallbacks& callbacks,
+                                 const Optional<std::chrono::milliseconds>& timeout)
+    : parent_(parent), stream_callbacks_(callbacks), stream_id_(parent.config_.random_.random()),
+      router_(parent.config_), request_info_(Protocol::Http11),
+      route_(parent_.cluster_.name(), timeout) {
+
+  router_.setDecoderFilterCallbacks(*this);
   // TODO: Correctly set protocol in request info when we support access logging.
 }
 
-AsyncRequestImpl::~AsyncRequestImpl() { ASSERT(!reset_callback_); }
+AsyncStreamImpl::~AsyncStreamImpl() { ASSERT(!reset_callback_); }
 
-void AsyncRequestImpl::encodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
-  response_.reset(new ResponseMessageImpl(std::move(headers)));
+void AsyncStreamImpl::encodeHeaders(HeaderMapPtr&& headers, bool end_stream) {
```

Comment:
I explained this above, this is an internal implementation detail, the response is being "encoded" into the client, which is then dispatching the appropriate callbacks.